### ISSUE TITLE
error: cannot find -lpython3.7

### ISSUE THREAD

**Issue by s1113950** [kivy/buildozer#842]
I initially filed this bug as https://github.com/kivy/python-for-android/issues/1728 and https://github.com/android-ndk/ndk/issues/918 (not sure which had the bug). More information is below.

### Versions

* Python: 3.7
* OS: Your Dockerfile in this repo (master branch version)
* Buildozer: master

### Description

I followed the buildozer steps off of master to build my android app with flask and a few other dependencies. I was able to build an ios app manually using kivy-ios but I figured I'd give buildozer a try, and then ran into the below:

### buildozer.spec
Command:
```sh
buildozer android debug deploy run
```

Spec file:
```
[app]

# (str) Title of your application
title = Piles

# (str) Package name
package.name = piles

# (str) Package domain (needed for android/ios packaging)
package.domain = piles.project

# (str) Source code where the main.py live
source.dir = .

# (list) Source files to include (let empty to include all the files)
source.include_exts = py,png,kv

# (list) List of inclusions using pattern matching
#source.include_patterns = assets/*,images/*.png

# (list) Source files to exclude (let empty to not exclude anything)
#source.exclude_exts = spec

# (list) List of directory to exclude (let empty to not exclude anything)
source.exclude_dirs = pong_game

# (list) List of exclusions using pattern matching
#source.exclude_patterns = license,images/*/*.jpg

# (str) Application versioning (method 1)
version = 0.1

# (str) Application versioning (method 2)
# version.regex = __version__ = ['"](.*)['"]
# version.filename = %(source.dir)s/main.py

# (list) Application requirements
# comma separated e.g. requirements = sqlite3,kivy
requirements = python3,kivy,flask,requests,https://github.com/pallets/jinja/zipball/master

# (str) Custom source folders for requirements
# Sets custom source for any requirements with recipes
# requirements.source.kivy = ../../kivy

# (list) Garden requirements
#garden_requirements =

# TODO: Have a presplash!
# (str) Presplash of the application
#presplash.filename = %(source.dir)s/data/presplash.png

# TODO: have an icon!
# (str) Icon of the application
#icon.filename = %(source.dir)s/data/icon.png

# (str) Supported orientation (one of landscape, sensorLandscape, portrait or all)
orientation = all

# (list) List of service to declare
#services = NAME:ENTRYPOINT_TO_PY,NAME2:ENTRYPOINT2_TO_PY

#
# OSX Specific
#

#
# author = © Copyright Info

# change the major version of python used by the app
osx.python_version = 3

# Kivy version to use
osx.kivy_version = 1.10.1

#
# Android specific
#

# (bool) Indicate if the application should be fullscreen or not
fullscreen = 1

# (string) Presplash background color (for new android toolchain)
# Supported formats are: #RRGGBB #AARRGGBB or one of the following names:
# red, blue, green, black, white, gray, cyan, magenta, yellow, lightgray,
# darkgray, grey, lightgrey, darkgrey, aqua, fuchsia, lime, maroon, navy,
# olive, purple, silver, teal.
#android.presplash_color = #FFFFFF

# (list) Permissions
#android.permissions = INTERNET,WAKE_LOCK

# (int) Target Android API, should be as high as possible.
#android.api = 27

# (int) Minimum API your APK will support.
#android.minapi = 21

# (int) Android SDK version to use
#android.sdk = 20

# (str) Android NDK version to use
#android.ndk = 17c

# (int) Android NDK API to use. This is the minimum API your app will support, it should usually match android.minapi.
#android.ndk_api = 21

# (bool) Use --private data storage (True) or --dir public storage (False)
#android.private_storage = True

# (str) Android NDK directory (if empty, it will be automatically downloaded.)
#android.ndk_path =

# (str) Android SDK directory (if empty, it will be automatically downloaded.)
#android.sdk_path =

# (str) ANT directory (if empty, it will be automatically downloaded.)
#android.ant_path =

# (bool) If True, then skip trying to update the Android sdk
# This can be useful to avoid excess Internet downloads or save time
# when an update is due and you just want to test/build your package
# android.skip_update = False

# (bool) If True, then automatically accept SDK license
# agreements. This is intended for automation only. If set to False,
# the default, you will be shown the license when first running
# buildozer.
# TODO: make this True
android.accept_sdk_license = True

# (str) Android entry point, default is ok for Kivy-based app
#android.entrypoint = org.renpy.android.PythonActivity

# (list) Pattern to whitelist for the whole project
#android.whitelist =

# (str) Path to a custom whitelist file
#android.whitelist_src =

# (str) Path to a custom blacklist file
#android.blacklist_src =

# (list) List of Java .jar files to add to the libs so that pyjnius can access
# their classes. Don't add jars that you do not need, since extra jars can slow
# down the build process. Allows wildcards matching, for example:
# OUYA-ODK/libs/*.jar
#android.add_jars = foo.jar,bar.jar,path/to/more/*.jar

# (list) List of Java files to add to the android project (can be java or a
# directory containing the files)
#android.add_src =

# (list) Android AAR archives to add (currently works only with sdl2_gradle
# bootstrap)
#android.add_aars =

# (list) Gradle dependencies to add (currently works only with sdl2_gradle
# bootstrap)
#android.gradle_dependencies =

# (list) Java classes to add as activities to the manifest.
#android.add_activites = com.example.ExampleActivity

# (str) python-for-android branch to use, defaults to master
#p4a.branch = master

# (str) OUYA Console category. Should be one of GAME or APP
# If you leave this blank, OUYA support will not be enabled
# android.ouya.category = GAME

# (str) Filename of OUYA Console icon. It must be a 732x412 png image.
#android.ouya.icon.filename = %(source.dir)s/data/ouya_icon.png

# (str) XML file to include as an intent filters in <activity> tag
#android.manifest.intent_filters =

# (str) launchMode to set for the main activity
#android.manifest.launch_mode = standard

# (list) Android additional libraries to copy into libs/armeabi
#android.add_libs_armeabi = libs/android/*.so
#android.add_libs_armeabi_v7a = libs/android-v7/*.so
#android.add_libs_x86 = libs/android-x86/*.so
#android.add_libs_mips = libs/android-mips/*.so

# (bool) Indicate whether the screen should stay on
# Don't forget to add the WAKE_LOCK permission if you set this to True
# TODO: do I wanna do this? :thinking_face:
android.wakelock = True

# (list) Android application meta-data to set (key=value format)
#android.meta_data =

# (list) Android library project to add (will be added in the
# project.properties automatically.)
#android.library_references =

# (str) Android logcat filters to use
#android.logcat_filters = *:S python:D

# (bool) Copy library instead of making a libpymodules.so
#android.copy_libs = 1

# (str) The Android arch to build for, choices: armeabi-v7a, arm64-v8a, x86
android.arch = armeabi-v7a

#
# Python for android (p4a) specific
#

# (str) python-for-android git clone directory (if empty, it will be automatically cloned from github)
p4a.source_dir = .buildozer/android/platform/python-for-android

# (str) The directory in which python-for-android should look for your own build recipes (if any)
#p4a.local_recipes =

# (str) Filename to the hook for p4a
#p4a.hook =

# (str) Bootstrap to use for android builds
# p4a.bootstrap = sdl2

# (int) port number to specify an explicit --port= p4a argument (eg for bootstrap flask)
#p4a.port =


#
# iOS specific
#

# (str) Path to a custom kivy-ios folder
#ios.kivy_ios_dir = ../kivy-ios
# Alternately, specify the URL and branch of a git checkout:
ios.kivy_ios_url = https://github.com/kivy/kivy-ios
ios.kivy_ios_branch = master

# Another platform dependency: ios-deploy
# Uncomment to use a custom checkout
#ios.ios_deploy_dir = ../ios_deploy
# Or specify URL and branch
ios.ios_deploy_url = https://github.com/phonegap/ios-deploy
ios.ios_deploy_branch = 1.9.4

# (str) Name of the certificate to use for signing the debug version
# Get a list of available identities: buildozer ios list_identities
#ios.codesign.debug = "iPhone Developer: <lastname> <firstname> (<hexstring>)"

# (str) Name of the certificate to use for signing the release version
#ios.codesign.release = %(ios.codesign.debug)s


[buildozer]

# (int) Log level (0 = error only, 1 = info, 2 = debug (with command output))
log_level = 2

# (int) Display warning if buildozer is run as root (0 = False, 1 = True)
warn_on_root = 1

# (str) Path to build artifact storage, absolute or relative to spec file
# build_dir = ./.buildozer

# (str) Path to build output (i.e. .apk, .ipa) storage
# bin_dir = ./bin

#    -----------------------------------------------------------------------------
#    List as sections
#
#    You can define all the "list" as [section:key].
#    Each line will be considered as a option to the list.
#    Let's take [app] / source.exclude_patterns.
#    Instead of doing:
#
#[app]
#source.exclude_patterns = license,data/audio/*.wav,data/images/original/*
#
#    This can be translated into:
#
#[app:source.exclude_patterns]
#license
#data/audio/*.wav
#data/images/original/*
#


#    -----------------------------------------------------------------------------
#    Profiles
#
#    You can extend section / key with a profile
#    For example, you want to deploy a demo version of your application without
#    HD content. You could first change the title to add "(demo)" in the name
#    and extend the excluded directories to remove the HD content.
#
#[app@demo]
#title = My Application (demo)
#
#[app:source.exclude_patterns@demo]
#images/hd/*
#
#    Then, invoke the command line with the "demo" profile:
#
#buildozer --profile demo android debug
```

### Logs

```
/home/user/.buildozer/android/platform/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../a$
m-linux-androideabi/bin/ld: error: cannot find -lpython3.7m
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:51: error: undefined reference to 'PyModule_
Create2'
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:158: error: undefined reference to 'Py_SetPr
ogramName'
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:164: error: undefined reference to 'PyImport
_AppendInittab'
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:197: error: undefined reference to 'Py_Decod
eLocale'
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:198: error: undefined reference to 'Py_SetPa
th'
/home/user/hostcwd/app/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:209: error: undefined reference to 'Py_Initi
alize'
...
```

The comment here https://github.com/android-ndk/ndk/issues/918#issuecomment-467928112 mentions a pre-built Python library. Does buildozer do this for python3.7 or do I have to do this myself and then link it somehow? Any help is appreciated!

**Comment by tshirtman** [comment]
I think the dockerfile is not up to date, i did a build from a clean ubuntu docker a few days ago with just these steps.

```
apt update
apt install -y \
  git zip unzip python3 python3-virtualenv python3-pip \
  openjdk-8-jdk pkg-config autoconf libtool zlib1g-dev

pip install cython buildozer
buildozer android debug
```

**Comment by AndreMiras** [comment]
Thanks for the bug report. If the Dockerfile is not up to date and building simple things from it fails, then when definitely have a bug. Building from Docker is not currently part of the continuous integration, so that's probably why we pay for it. I have a couple of other tasks to address first, but this one is definitely on my TODO list, so stay tuned @s1113950 !
Also first thing is obviously I'll also try with Docker on my local to see if at least I can reproduce. But in any case at the end it should be part of the CI, so you can see it working with your on eyes :smile:

**Comment by s1113950** [comment]
I'll run with the simpler dockerfile @tshirtman ! I made a few tweaks to the dockerfile off of `master` to enable running with `master` of `buildozer` rather than the old version on `pip`. I'll give the older version a try and see if I have better luck.

Here's a diff of my tweaks:
```
diff --git a/Dockerfile b/Dockerfile
index 36dacde..5fe698d 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -15,15 +15,13 @@
 FROM ubuntu:18.04
 ARG git

-ENV BDOZER_REQ=${git:+"/src"}
-ENV BDOZER_REQ=${BDOZER_REQ:-"buildozer"}
+ARG BDOZER_REQ=${git:+"/src"}
+ARG BDOZER_REQ=${BDOZER_REQ:-"buildozer"}
 ENV USER="user"
 ENV HOME_DIR="/home/${USER}"
 ENV WORK_DIR="${HOME_DIR}/hostcwd" \
     PATH="${HOME_DIR}/.local/bin:${PATH}"

-COPY . /src
-
 # configures locale
 RUN apt update -qq > /dev/null && \
     apt install -qq --yes --no-install-recommends \
@@ -50,17 +48,20 @@ RUN useradd --create-home --shell /bin/bash ${USER}
 RUN usermod -append --groups sudo ${USER}
 RUN echo "%sudo ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

+COPY . /src
+
 USER ${USER}
 WORKDIR ${WORK_DIR}

 # installs buildozer and dependencies
-RUN pip install --user Cython==0.25.2 $BDOZER_REQ
+RUN pip install --user Cython==0.25.2 /src

 # calling buildozer adb command should trigger SDK/NDK first install and update
 # but it requires a buildozer.spec file
-RUN cd /tmp/ && buildozer init && buildozer android adb -- version && cd -
+RUN cd /tmp/ && buildozer init && (yes | buildozer android adb -- version) && cd -
 # fixes source and target JDK version, refs https://github.com/kivy/buildozer/issues/625
-RUN sed s/'name="java.source" value="1.5"'/'name="java.source" value="7"'/ -i ${HOME_DIR}/.buildozer/android/platform/android-sdk-20/tools/ant/build.xml
-RUN sed s/'name="java.target" value="1.5"'/'name="java.target" value="7"'/ -i ${HOME_DIR}/.buildozer/android/platform/android-sdk-20/tools/ant/build.xml
+# RUN sed s/'name="java.source" value="1.5"'/'name="java.source" value="7"'/ -i ${HOME_DIR}/.buildozer/android/platform/android-sdk-20/tools/ant/build.xml
+# RUN sed s/'name="java.target" value="1.5"'/'name="java.target" value="7"'/ -i ${HOME_DIR}/.buildozer/android/platform/android-sdk-20/tools/ant/build.xml
```
The `sed` replacement also didn't work for me so I removed it. There was no `android-sdk-20`, it was a different path:
```
user@d5026e3b1d1a:~/hostcwd$ ls /home/user/.buildozer/android/platform/
android  android-ndk-r17c  android-ndk-r17c-linux-x86_64.zip  android-sdk  apache-ant-1.9.4  apache-ant-1.9.4-bin.tar.gz
user@d5026e3b1d1a:~/hostcwd$
```
The `android-sdk/tools` folder didn't have `ant`:
```
user@d5026e3b1d1a:~/hostcwd$ ls /home/user/.buildozer/android/platform/android-sdk/tools
android  bin  emulator  emulator-check  lib  mksdcard  monitor  NOTICE.txt  package.xml  proguard  source.properties  support
```
I also had to use `yes | ...` to get around the license request when running `buildozer android adb` for the first time. 

Finally, I modified the `log_level` of `buildozer/default.spec` to be `2` from `1` so the prompt would appear to accept the license agreement and `yes` would work.

If you guys want I can make a PR with these changes, but most of them seemed hacky to me so I didn't think it was a good enough fix (and because I ran into so many other issues I wasn't sure if it was just my system or not), with the exception of enabling the current codebase to be built in the dockerfile rather than `buildozer` from a certain version.

**Comment by s1113950** [comment]
I used this Dockerfile @tshirtman (only difference being pip installing virtualenv rather than from apt so as to not get an error saying that there's no virtualenv in the path):
```
FROM ubuntu:18.04

RUN apt-get update -y && apt install -y \
  git \
  zip \
  unzip \
  python3 \
  python3-pip \
  openjdk-8-jdk \
  pkg-config \
  autoconf \
  libtool \
  zlib1g-dev

RUN pip3 install cython buildozer virtualenv

# so can debug and run buildozer android deploy debug run
ENTRYPOINT ["/bin/bash"]
```
, and then inside of that I ran `buildozer android debug deploy run`, and here's the error:
```
Building sdl2 for armeabi-v7a
[INFO]:    -> directory context /home/user/hostcwd/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni
[INFO]:    -> running ndk-build V=1
...
make: *** No rule to make target `/home/user/.buildozer/android/platform/android-ndk-r17c/sources/android/cpufeatures/cpu-features.c', needed by `/home/user/hostcwd/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/objs/cpufeatures/cpu-features.o'.  Stop.


  STDERR:

# Command failed: /usr/bin/python3 -m pythonforandroid.toolchain create --dist_name=piles --bootstrap=sdl2 --requirements=python3,kivy,flask,requests,https://github.com/pallets/jinja/zipball/master --arch armeabi-v7a --copy-libs --color=always --storage-dir="/home/user/hostcwd/.buildozer/android/platform/build" --ndk-api=21
#
# Buildozer failed to execute the last command
# The error might be hidden in the log above this error
# Please read the full log, and search for it before
# raising an issue with buildozer itself.
# In case of a bug report, please add a full log with log_level = 2

😞 I'll play with it but I'm not really sure why `sdl2` won't build; it's a bit of a "black box" to me unfortunately.

**Comment by s1113950** [comment]
https://stackoverflow.com/a/30525109/7111774 might be of some help?  Perhaps `android-ndk-r17c` isn't in the expected place?

**Comment by s1113950** [comment]
Interesting! So initially I ran as `root` despite the warning it gave, and I got the error I posted. When I modified the dockerfile to add a non-root user, I get the error I got previously: 
```
/home/user/.buildozer/android/platform/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../ar
m-linux-androideabi/bin/ld: error: cannot find -lpython3.7m
```

**Comment by s1113950** [comment]
Nothing changed on my end, but now I can't get past this step:
```
[INFO]:    Will compile for the following archs: armeabi-v7a
[INFO]:    Found Android API target in $ANDROIDAPI: 27
[INFO]:    Available Android APIs are ()
[ERROR]:   Build failed: Requested API target 27 is not available, install it with the SDK android tool.
# Command failed: /usr/bin/python3 -m pythonforandroid.toolchain create --dist_name=piles --bootstrap=sdl2 --requirements=python3,kivy,flask,requests,https://github.com/pallets/jinja/zipball/master --arch armeabi-v7a --copy-libs --color=always --storage-dir="/data/.buildozer/android/platform/build" --ndk-api=21
#
# Buildozer failed to execute the last command
# The error might be hidden in the log above this error
# Please read the full log, and search for it before
# raising an issue with buildozer itself.
# In case of a bug report, please add a full log with log_level = 2
```
I tried to check if some code was just merged into the android-ndk but there's no code here: https://github.com/android-ndk/ndk 🤔

**Comment by s1113950** [comment]
Sorry for so many comments :( Last thing I have though: I was able to manually build my app using `python-for-android` and the `Dockerfile.py3` that was located there, along with this run command:
```
docker run -e ANDROIDSDK="/opt/android/android-sdk" -e ANDROIDNDK="/opt/android/android-ndk-r17c" -e ANDROIDAPI="27" -e NDKAPI="21" --volume myapp:/app -it --rm p4a /bin/bash
```
And this build command: 
``` 
. venv/bin/activate && p4a apk --private /app --package=org.app.app --name "App" --version 0.1 --bootstrap=sdl2 --requirements=python3,kivy,flask,requests,https://github.com/pallets/jinja/zipball/master
```

**Comment by AndreMiras** [comment]
Yes the Dockerfile provided on p4a is working and part of the integration testing.
Nonetheless I still want us to also provide one that's useful for buildozer. But we need to make it part of the integration testing before tweaking it further. I also want us to auto-build it on docker-hub just like the p4a one. That's currently an undocumented p4a feature, but you could have used the pre-built image with building it `docker pull kivy/python-for-android`

**Comment by Shu-Ji** [comment]
the same erro on OSX 10.13.6:
```
$ uname -a
Darwin rmbp-finn.lan 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
```

**Comment by AndreMiras** [comment]
We published a new Docker image on DockerHub which seems to work for compiling your project. See https://hub.docker.com/r/kivy/buildozer
I gave it a try with:
```
docker run --volume "$HOME/.buildozer":/home/user/.buildozer --volume "$PWD":/home/user/hostcwd kivy/buildozer android debug
```
And it could compile just fine, see log extract:
```
[DEBUG]:        :createDebugCompatibleScreenManifests
[DEBUG]:        :processDebugManifest
[DEBUG]:        :splitsDiscoveryTaskDebug
[DEBUG]:        :processDebugResources
[DEBUG]:        :generateDebugSources
[DEBUG]:        :javaPreCompileDebug
[DEBUG]:        :compileDebugJavaWithJavacNote: Some input files use or override a deprecated API.
[DEBUG]:        Note: Recompile with -Xlint:deprecation for details.
[DEBUG]:        Note: Some input files use unchecked or unsafe operations.
[DEBUG]:        Note: Recompile with -Xlint:unchecked for details.
[DEBUG]:
[DEBUG]:        :compileDebugNdk NO-SOURCE
[DEBUG]:        :compileDebugSources
[DEBUG]:        :mergeDebugShaders
[DEBUG]:        :compileDebugShaders
[DEBUG]:        :generateDebugAssets
[DEBUG]:        :mergeDebugAssets
[DEBUG]:        :transformClassesWithDexBuilderForDebug
[DEBUG]:        :transformDexArchiveWithExternalLibsDexMergerForDebug
[DEBUG]:        :transformDexArchiveWithDexMergerForDebug
[DEBUG]:        :mergeDebugJniLibFolders
[DEBUG]:        :transformNativeLibsWithMergeJniLibsForDebug
[DEBUG]:        :transformNativeLibsWithStripDebugSymbolForDebug
[DEBUG]:        :processDebugJavaRes NO-SOURCE
[DEBUG]:        :transformResourcesWithMergeJavaResForDebug
[DEBUG]:        :validateSigningDebug
[DEBUG]:        :packageDebug
[DEBUG]:        :assembleDebug
[DEBUG]:
[DEBUG]:        BUILD SUCCESSFUL in 1m 8s
[DEBUG]:        27 actionable tasks: 27 executed
[INFO]:    <- directory context /home/user/hostcwd/.buildozer/android/platform/python-for-android
[INFO]:    # Copying APK to current directory
[INFO]:    # APK filename not found in build output. Guessing...
[INFO]:    # Found APK file: /home/user/hostcwd/.buildozer/android/platform/build/dists/piles/build/outputs/apk/debug/piles-debug.apk
[INFO]:    # Add version number to APK
[INFO]:    # APK renamed to piles-0.1-debug.apk
stty: 'standard input': Inappropriate ioctl for device
[DEBUG]:   -> running cp /home/user/hostcwd/.buildozer/android/platform/build/dists/piles/build/outputs/apk/debug/piles-debug.apk piles-0.1-debug.apk
WARNING: Received a --sdk argument, but this argument is deprecated and does nothing.
No compiled python is present to zip, skipping.
Applying Java source code patches...
Applying patch: src/patches/SDLActivity.java.patch
# Android packaging done!
# APK piles-0.1-debug.apk available in the bin directory
```
Currently the docker image is an undocumented feature on purpose because we're still figuring out if it does provide value and if/how we could improve it.
I'm closing the issue for now because I got it working for your specific case. Feel free to come back with feedback on this task or Discord if you have struggles related to Docker

**Comment by FunmiKesa** [comment]
> the same erro on OSX 10.13.6 without Dockerfile:
> 
> ```
> $ uname -a
> Darwin rmbp-finn.lan 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
> ```
> I don't find the `libpython3.7m.so` in the `.buildozer` directory.
> But find `libpython3.7m.a` in:
> 
> ```
> .buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/libpython3.7m.a
> ```

 I currently have the same issue. How did you solve this?

**Comment by mgc8** [comment]
I'm also experiencing the same issue, using macOS 10.14.3 and no Docker -- just plain Buildozer from the PyPi repo. What is the recommended course of action to fix this?

`/.buildozer/android/platform/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpython3.7m`

**Comment by bsnresearcher** [comment]
I met same problem on macos and virtual machine. However, I solved by having built on where was not shared file system of virtual machine (maybe, including virtualbox or docker too).  I could not have solved to build on mac os yet.

Python-for-android seems to be weak to aquire path infomation required to build apk.  I hope it will be improved.

**Comment by bsnresearcher** [comment]
I met same problem on macos and virtual machine. 

However, I solved by having built on where was not shared file system of virtual machine (maybe, including virtual-box or docker too).  Not have needed to Install  python3.7-dev and libpython3.7-dev and so on.

I could not have solved to build on macos yet.

**Comment by jamesdeluk** [comment]
I also have this problem.

macOS 10.14.6, buildozer 0.39, Python 3.7.4

Very simply program - only an input, button, and label. Only the basics (title, package, etc) edited in .init file.

`[armeabi-v7a] SharedLibrary  : libmain.so
/Users/James/.buildozer/android/platform/android-ndk-r17c/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ -Wl,-soname,libmain.so -shared --sysroot=/Users/James/.buildozer/android/platform/android-ndk-r17c/platforms/android-21/arch-arm /Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/objs/main/__/__/SDL/src/main/android/SDL_android_main.o /Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/objs/main/start.o -lgcc -Wl,--exclude-libs,libgcc.a -latomic -Wl,--exclude-libs,libatomic.a /Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/libSDL2.so  -gcc-toolchain /Users/James/.buildozer/android/platform/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64 -no-canonical-prefixes -target armv7-none-linux-androideabi21 -Wl,--fix-cortex-a8 -L/Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/other_builds/python3-libffi-openssl-sqlite3/armeabi-v7a__ndk_target_21/python3/android-build -Wl,--build-id -Wl,--no-undefined -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--warn-shared-textrel -Wl,--fatal-warnings   -L/Users/James/.buildozer/android/platform/android-ndk-r17c/platforms/android-21/arch-arm/usr/lib -lGLESv1_CM -lGLESv2 -llog -lpython3.7m -lstdc++ -lc -lm -o /Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/libmain.so
/Users/James/.buildozer/android/platform/android-ndk-r17c/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpython3.7m
/Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:47: error: undefined reference to 'PyModule_Create2'
/// many more
/Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/jni/application/src/start.c:0: error: undefined reference to '_Py_NoneStruct'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/Users/James/Downloads/CheckerApp/.buildozer/android/platform/build/build/bootstrap_builds/sdl2-python3/obj/local/armeabi-v7a/libmain.so] Error 1

  STDERR:

# Command failed: /usr/local/opt/python/bin/python3.7 -m pythonforandroid.toolchain create --dist_name=aswpc --bootstrap=sdl2 --requirements=python3,kivy --arch armeabi-v7a --copy-libs --color=always --storage-dir="/Users/James/Downloads/CheckerApp/.buildozer/android/platform/build" --ndk-api=21`

**Comment by Shu-Ji** [comment]
> > the same erro on OSX 10.13.6 without Dockerfile:
> > ```
> > $ uname -a
> > Darwin rmbp-finn.lan 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
> > ```
> > 
> > 
> > I don't find the `libpython3.7m.so` in the `.buildozer` directory.
> > But find `libpython3.7m.a` in:
> > ```
> > .buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/libpython3.7m.a
> > ```
> 
> I currently have the same issue. How did you solve this?



> > the same erro on OSX 10.13.6 without Dockerfile:
> > ```
> > $ uname -a
> > Darwin rmbp-finn.lan 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
> > ```
> > 
> > 
> > I don't find the `libpython3.7m.so` in the `.buildozer` directory.
> > But find `libpython3.7m.a` in:
> > ```
> > .buildozer/android/platform/build/build/other_builds/hostpython3/desktop/hostpython3/native-build/libpython3.7m.a
> > ```
> 
> I currently have the same issue. How did you solve this?

Sorry for the answer, I haven't resolved it after 6 months later.
I have head to Flutter.
Hope someone can solve this.

**Comment by bsnresearcher** [comment]
I solved to have built my app on Ubuntu in virtual-machine (e.g. docker. I used VMWare Fusion11) instead of MacOS.

buildozer calls [python-for-android](https://github.com/kivy/python-for-android)  inside, which package seems not to fail building on Ubuntu more more than on MacOS. On the other hand, it failed on MacOS frequently (Because of insufficient test on MacOS(?)).

So your source files may pass building on Ubuntu. 

Caution) you must not build your app on sharing file system on VM. It fails to build.
